File: go/core/metadata_details_test.go
package core

import (
	"context"
	"encoding/json"
	"strings"
	"testing"
)

// --- LLM request metadata (Rust parity: ChatRequest.metadata / with_metadata) ---

func TestMetadataAbsentByDefault(t *testing.T) {
	mock := NewMockLlmProvider().PushText("ok")
	agent := NewSmoothAgent(mock, AgentOptions{})
	if _, err := agent.Run(context.Background(), "hi", nil); err != nil {
		t.Fatal(err)
	}
	last, ok := mock.LastCall()
	if !ok {
		t.Fatal("no recorded call")
	}
	if last.Metadata != nil {
		t.Fatalf("metadata must be nil by default, got %v", last.Metadata)
	}
}

func TestMetadataForwardedVerbatim(t *testing.T) {
	mock := NewMockLlmProvider().PushText("ok")
	meta := map[string]any{"smooai_agent_slug": "support-bot", "k": "v"}
	agent := NewSmoothAgent(mock, AgentOptions{Metadata: meta})
	if _, err := agent.Run(context.Background(), "hi", nil); err != nil {
		t.Fatal(err)
	}
	last, _ := mock.LastCall()
	if last.Metadata["smooai_agent_slug"] != "support-bot" || last.Metadata["k"] != "v" {
		t.Fatalf("metadata must be forwarded verbatim, got %v", last.Metadata)
	}
}

func TestEmptyMetadataNormalizedToNil(t *testing.T) {
	mock := NewMockLlmProvider().PushText("ok")
	agent := NewSmoothAgent(mock, AgentOptions{Metadata: map[string]any{}})
	if _, err := agent.Run(context.Background(), "hi", nil); err != nil {
		t.Fatal(err)
	}
	last, _ := mock.LastCall()
	if last.Metadata != nil {
		t.Fatalf("empty metadata must normalize to nil (wire-identical to unset), got %v", last.Metadata)
	}
}

func TestWireRequestMetadataOmittedAndForwarded(t *testing.T) {
	// nil metadata → no "metadata" key on the wire at all.
	bare, err := json.Marshal(buildWireRequest(ChatRequest{Model: "m"}, false))
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(bare), "metadata") {
		t.Fatalf("unset metadata must be omitted from the wire: %s", bare)
	}
	// empty metadata → normalized away, byte-identical to unset.
	empty, err := json.Marshal(buildWireRequest(ChatRequest{Model: "m", Metadata: map[string]any{}}, false))
	if err != nil {
		t.Fatal(err)
	}
	if string(empty) != string(bare) {
		t.Fatalf("empty metadata must be wire-identical to unset:\n%s\n%s", empty, bare)
	}
	// set metadata → forwarded verbatim as a top-level object.
	tagged, err := json.Marshal(buildWireRequest(ChatRequest{Model: "m", Metadata: map[string]any{"k": "v"}}, false))
	if err != nil {
		t.Fatal(err)
	}
	var parsed map[string]any
	if err := json.Unmarshal(tagged, &parsed); err != nil {
		t.Fatal(err)
	}
	md, ok := parsed["metadata"].(map[string]any)
	if !ok || md["k"] != "v" {
		t.Fatalf("metadata must be forwarded verbatim: %s", tagged)
	}
}

// --- Structured tool details (Rust parity: ToolResult.details → ToolCallComplete.details) ---

// detailsHook attaches structured details in PostCall, the same seam the Rust
// engine populates ToolResult.details through.
type detailsHook struct{ details any }

func (h detailsHook) PreCall(context.Context, ToolCall) error { return nil }
func (h detailsHook) PostCall(_ context.Context, _ ToolCall, result *ToolResult) error {
	result.Details = h.details
	return nil
}

func TestStreamToolResultForwardsDetailsFromHook(t *testing.T) {
	mock := NewMockLlmProvider().
		PushToolCall("call_1", "echo", `{"text":"hi"}`).
		PushText("done")
	details := map[string]any{"traceId": "abc123", "errorCount": 47}
	agent := NewSmoothAgent(mock, AgentOptions{
		Tools: []Tool{FuncTool{
			ToolName: "echo",
			Desc:     "echoes",
			Params:   map[string]any{"type": "object"},
			Fn: func(_ context.Context, args map[string]any) (string, error) {
				s, _ := args["text"].(string)
				return s, nil
			},
		}},
		Hooks: []ToolHook{detailsHook{details: details}},
	})

	stream, err := agent.RunStream(context.Background(), "go", nil)
	if err != nil {
		t.Fatal(err)
	}
	events, err := drain(stream)
	if err != nil {
		t.Fatal(err)
	}
	found := false
	for _, e := range events {
		if e.Kind == StreamToolResult {
			found = true
			d, ok := e.Details.(map[string]any)
			if !ok || d["traceId"] != "abc123" {
				t.Fatalf("details must be forwarded verbatim on StreamToolResult, got %v", e.Details)
			}
		}
	}
	if !found {
		t.Fatal("expected a StreamToolResult event")
	}
}

func TestStreamToolResultDetailsNilWithoutHook(t *testing.T) {
	mock := NewMockLlmProvider().
		PushToolCall("call_1", "echo", `{"text":"hi"}`).
		PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{
		Tools: []Tool{FuncTool{
			ToolName: "echo",
			Desc:     "echoes",
			Params:   map[string]any{"type": "object"},
			Fn: func(_ context.Context, args map[string]any) (string, error) {
				s, _ := args["text"].(string)
				return s, nil
			},
		}},
	})

	stream, err := agent.RunStream(context.Background(), "go", nil)
	if err != nil {
		t.Fatal(err)
	}
	events, err := drain(stream)
	if err != nil {
		t.Fatal(err)
	}
	for _, e := range events {
		if e.Kind == StreamToolResult && e.Details != nil {
			t.Fatalf("details must be nil when no hook attaches them, got %v", e.Details)
		}
	}
}

File: typescript/core/test/metadataDetails.test.ts
/**
 * LLM request metadata + structured tool details — the TypeScript port of the
 * Rust reference's `ChatRequest.metadata` / `with_metadata` (LiteLLM spend-log
 * attribution) and `AgentEvent::ToolCallComplete.details` (UI-facing structured
 * tool payload attached by a postCall hook).
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, metadataField, SmoothAgent, StreamEvent, Tool, ToolHook } from '../src/agent.js';
import { MockLlmProvider } from '../src/llmProvider.js';

async function collect(gen: AsyncGenerator<StreamEvent>): Promise<StreamEvent[]> {
    const events: StreamEvent[] = [];
    for await (const e of gen) events.push(e);
    return events;
}

const echo: Tool = {
    name: 'echo',
    description: 'Echoes input',
    parameters: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] },
    execute: async (args) => `echoed:${String(args.text ?? '')}`,
};

describe('LLM request metadata', () => {
    it('is absent from the request body by default', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('ok');
        const agent = new SmoothAgent(mock, {} satisfies AgentOptions);
        await agent.run('hi');
        expect('metadata' in mock.calls[0].body).toBe(false);
    });

    it('is forwarded verbatim as the top-level metadata object', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('ok');
        const agent = new SmoothAgent(mock, { metadata: { smooai_agent_slug: 'support-bot', k: 'v' } } satisfies AgentOptions);
        await agent.run('hi');
        expect(mock.calls[0].body.metadata).toEqual({ smooai_agent_slug: 'support-bot', k: 'v' });
    });

    it('an empty metadata object is wire-identical to unset', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('ok');
        const agent = new SmoothAgent(mock, { metadata: {} } satisfies AgentOptions);
        await agent.run('hi');
        expect('metadata' in mock.calls[0].body).toBe(false);
    });

    it('metadataField normalizes empty/absent to {}', () => {
        expect(metadataField(undefined)).toEqual({});
        expect(metadataField({})).toEqual({});
        expect(metadataField({ k: 'v' })).toEqual({ metadata: { k: 'v' } });
    });
});

describe('structured tool details', () => {
    it('forwards details attached by a postCall hook on the tool_result event', async () => {
        const details = { traceId: 'abc123', errorCount: 47 };
        const hook: ToolHook = {
            postCall: async (_call, result) => {
                result.details = details;
            },
        };
        const mock = new MockLlmProvider();
        mock.pushToolCall('call-1', 'echo', '{"text":"ping"}');
        mock.pushText('done');

        const agent = new SmoothAgent(mock, { tools: [echo], toolHooks: [hook] } satisfies AgentOptions);
        const events = await collect(agent.runStream('use echo'));

        const toolResult = events.find((e) => e.type === 'tool_result') as { result: string; details?: unknown } | undefined;
        expect(toolResult?.result).toBe('echoed:ping');
        expect(toolResult?.details).toEqual(details);
    });

    it('details is undefined when no hook attaches it', async () => {
        const mock = new MockLlmProvider();
        mock.pushToolCall('call-1', 'echo', '{"text":"ping"}');
        mock.pushText('done');

        const agent = new SmoothAgent(mock, { tools: [echo] } satisfies AgentOptions);
        const events = await collect(agent.runStream('use echo'));

        const toolResult = events.find((e) => e.type === 'tool_result') as { details?: unknown } | undefined;
        expect(toolResult?.details).toBeUndefined();
    });
});
